CMake configuration

atteneder opened the pull request:
Created a CMake configuration.

Purpose: Make it easy to integrate KTX loading/transcoding in 3rd party projects (especially CMake based ones)

It's a starting point to fix #44

It's the 2nd revision (first was PR #124 ). Improvements:
- Less changes in existing source code
- Not mixed with other features (C bindings, Unity library)
- vkloadtest is running
- GitHub Actions to build for macOS and Linux

## Status

- [x] build ktx library
  - [ ] compiler flag parity with gyp
  - [ ] Javascript bindings
- [ ] Documentation
- [x] CTest testing
  - [x] unittests
  - [x] texturetests
  - [x] transcodetests
- [x] tools
  - [x] ktx2check
  - [x] ktx2ktx2
  - [x] ktxinfo
  - [x] ktxsc
  - [x] toktx
- [x] Load test apps
  - [x] glloadtest
  - [x] vkloadtest

## Tested on platforms 
- [x] macOS
- [ ] iOS
- [ ] Android
- [ ] Web (via emscripten)
- [x] Linux
- [ ] Windows

## CI

- [ ] Appveyor
- [ ] Travis CI
- [ ] Documentation GitHub pages deployment
- [x] GitHub actions

MarkCallow commented:
You can compile this for win32? I removed win32 support from the ktx2 branch because of the 64-bit offsets and (some) sizes in .ktx2. I thought it would be a nightmare trying to get it all to compile for a 32-bit platform without warnings.

atteneder commented:
Well, partially...and not without warnings.
What I did was compile my `ktx_unity`target in [another branch](https://github.com/atteneder/KTX-Software/tree/cmake_unity_basisu), which contains only reading and transcoding.
And then I back-ported the CMake changes that were necessary. I can try to build the full library for win32 in the next days, if you're interested.

MarkCallow commented:
> I can try to build the full library for win32 in the next days, if you're interested.

I don't think that is necessary. 32-bit systems are becoming increasingly rare and Apple has completely dropped 32-bit support.

MarkCallow commented on CMakeLists.txt:
```diff
@@ -0,0 +1,366 @@
+
+if(IOS)
+    set(CMAKE_OSX_DEPLOYMENT_TARGET "10.0" CACHE STRING "Minimum iOS deployment version")
+endif()
```
The equivalent Xcode setting is called `IPHONEOS_DEPLOYMENT_TARGET`. I hope this is a typo here and not a stupid design decision in CMake.

There also needs to be an

```
if (macOS)
    set(CMAKE_OSX_DEPLOYMENT_TARGET "10.11" CACHE STRING "Minimum macOS deployment version")
endif()
```

And please change "version" to "target" in those strings.

atteneder replied:
Well, `CMAKE_OSX_DEPLOYMENT_TARGET` can be used to set the deployment target for any Apple platforms (including macOS, iOS, watchOS, etc.).
I'll make sure there's a distinct default deployment target for both macOS and iOS.

MarkCallow commented:
You will need to rebase and modify this PR in light of PR #146. The message on the corresponding commit, https://github.com/KhronosGroup/KTX-Software/commit/1ec5d917c68f274abf68aa27f36844a5b72edc65, should give you a reasonable idea of what needs to be done.

atteneder commented:
I rebased it and added the new transcodetest.
It became clear to me that this CMake configuration does not use the same compiler warning flags as the current build system. I put it on the TODO list

MarkCallow commented:
Thanks. I made another small change to the flags, for Linux only. It now sets `-Wno-pedantic` only for c++ files as the anon structs and unions are only in c++ files.

MarkCallow commented:
Please rebase this against the latest ktx2. I've added JS bindings which necessitated some build changes.

I think I will soon be ready to switch over to cmake.

atteneder commented:
I rebased and build+tests+installs work again.
Feel free to have a look.

Also: What is missing to consider switching?
thanks

MarkCallow commented:
Thank you very much.

Can you build with Emscripten and can you build the JS bindings? That is the stuff built by build/cmake/web in the current build system. I ask because the Emscripten/web build of the libktx and ktxtests is fairly new and the JS bindings and their builds very new.

I need to add mipmap generation to toktx somewhat urgently. Once that is done I want to switch to the cmake build system and then I want to merge everything to master.

MarkCallow commented:
> What is missing to consider switching?

- [ ] Build of `libktx` and `ktxtests` targets with Emscripten for the "web" platform.
- [ ] Building of the wrappers in `interface/js_binding` when platform is "web". Targets are `libktx_js` and `msc_basis_transcoder_js` in the current build system.
- [ ] Fixing the Appveyor and Travis CI builds to use the new CMake based build. I do not understand the relationship between GitHub workflows and these existing CI builds. I searched the GitHub documentation and failed to find any information, not helped by the massive conflict between general GitHub workflows and the thing called GitHub Workflows. We need working CI builds on at least the current platforms (Windows with VC2015, 2017 and 2019, Linux, iOS, macOS and Web). On Windows, macOS and Linux, builds need to include the docs and running `toktx_tests` and `ktx2check_tests`. On macOS it needs to include building `package.tools`. One of the builds, currently I think macOS is doing this, needs to deploy the docs to GitHub pages. Maybe this is all already working with the workflows/main.yml that this PR adds. Where do I find the results of those builds?
- [ ] Addressing my feedback w.r.t `DEPLOYMENT_TARGET` for macOS and iOS.
- [ ] Updating BUILDING.md.

Nice to have are:

- [ ] CI build for Android
- [ ] Deployment of Mac installation package (`package.tools` target) to somewhere people can easily download it and which doesn't conflict with the docs deployment which completely rewrites GitHub pages.

MarkCallow commented on tests/loadtests/appfwSDL/VulkanAppSDL/VulkanAppSDL.cpp:
```diff
@@ -713,7 +713,7 @@ VulkanAppSDL::createDevice()
                                                NULL,
                                                &deviceExtensionCount,
-                                               NULL);
+                                               (vk::ExtensionProperties*)NULL);
```
Why not use `nullptr` here like you did at line 601?

atteneder replied:
FIxed!
I guess I oriented myself on two line above, but the whole file is a mixture of NULL vs. nullptr.

MarkCallow replied:
Yeah. It started out as all C code. I started converting to C++ (vulkan.hpp) but I've never had time to finish the job, I still hope to some day.

MarkCallow commented:
@atteneder I have changed the spec. for mip level padding and updated all the software so I'm afraid another rebase is needed. There were  no changes to the build, only the software.

I would like to complete adding mipmap generation to toktx and then merge ktx2 to master. I could do the merge first to give you more time to cope with that rebase. Which do you prefer. (Note that along with mipmap generation some new load tests will be added which the build system will have to be updated to know about.) I do not want to wait to do the merge until after the build system switch because several people have asked for it to be done soon and I don't know how long completing the build system switch will take.

atteneder commented:
> @atteneder I have changed the spec. for mip level padding and updated all the software so I'm afraid another rebase is needed. There were no changes to the build, only the software.
> 
> I would like to complete adding mipmap generation to toktx and then merge ktx2 to master. I could do the merge first to give you more time to cope with that rebase. Which do you prefer. (Note that along with mipmap generation some new load tests will be added which the build system will have to be updated to know about.) I do not want to wait to do the merge until after the build system switch because several people have asked for it to be done soon and I don't know how long completing the build system switch will take.

Actually I'm also eagerly waiting for ktx2 to be done, so I'm totally fine with full force on finishing it! :)
So don't worry about me rebasing this PR. I might not do it right away (since I'm doing this in leisure time), but it'll work out.

Thanks!

MarkCallow commented:
Please revert https://github.com/KhronosGroup/KTX-Software/pull/134/commits/63ac8fb80eb29542c36ca0334c1015deae6632d7. version.h files are generated during the build from Git information by the mkversion script.
